src/iterators.jl

```diff
@@ -0,0 +1,173 @@
+    # define a method for increment! This way we avoid cases where one may wish to nest
+    # calls to different step! methods accidentaly incrementing multiple times.
+    init || increment!(iterator)
+    rv = compute!(iterator)
+    return rv, false
+end
+
+function increment! end
+compute!(iterator::AbstractNetworkIterator) = iterator
+
+step!(iterator::AbstractNetworkIterator) = step!(identity, iterator)
+function step!(f, iterator::AbstractNetworkIterator)
+    compute!(iterator)
+    f(iterator)
+    increment!(iterator)
+    return iterator
+end
+
+#
+# RegionIterator
+#
+"""
+  struct RegionIterator{Problem, RegionPlan} <: AbstractNetworkIterator
+"""
+mutable struct RegionIterator{Problem, RegionPlan} <: AbstractNetworkIterator
+    problem::Problem
+    region_plan::RegionPlan
+    which_region::Int
+    const which_sweep::Int
+    function RegionIterator(problem::P, region_plan::R, sweep::Int) where {P, R}
+        if length(region_plan) == 0
+            throw(BoundsError("Cannot construct a region iterator with 0 elements."))
+        end
+        return new{P, R}(problem, region_plan, 1, sweep)
+    end
+end
+
+function RegionIterator(problem; sweep, sweep_kwargs...)
+    plan = region_plan(problem; sweep_kwargs...)
+    return RegionIterator(problem, plan, sweep)
+end
+
+function new_region_iterator(iterator::RegionIterator; sweep_kwargs...)
+    return RegionIterator(iterator.problem; sweep_kwargs...)
+end
+
+state(region_iter::RegionIterator) = region_iter.which_region
+Base.length(region_iter::RegionIterator) = length(region_iter.region_plan)
+
+problem(region_iter::RegionIterator) = region_iter.problem
+
+function current_region_plan(region_iter::RegionIterator)
+    return region_iter.region_plan[region_iter.which_region]
+end
+
+function current_region(region_iter::RegionIterator)
+    region, _ = current_region_plan(region_iter)
+    return region
+end
+
+function region_kwargs(region_iter::RegionIterator)
+    _, kwargs = current_region_plan(region_iter)
+    return kwargs
+end
+function region_kwargs(f::Function, iter::RegionIterator)
+    return get(region_kwargs(iter), Symbol(f, :_kwargs), (;))
+end
+
+function prev_region(region_iter::RegionIterator)
+    state(region_iter) <= 1 && return nothing
+    prev, _ = region_iter.region_plan[region_iter.which_region - 1]
+    return prev
+end
+
+function next_region(region_iter::RegionIterator)
+    islaststep(region_iter) && return nothing
+    next, _ = region_iter.region_plan[region_iter.which_region + 1]
+    return next
+end
+
+#
+# Functions associated with RegionIterator
+#
+function increment!(region_iter::RegionIterator)
+    region_iter.which_region += 1
+    return region_iter
+end
+
+function compute!(iter::RegionIterator)
+    _, local_state = extract!(iter; region_kwargs(extract!, iter)...)
+    _, local_state = update!(iter, local_state; region_kwargs(update!, iter)...)
+    insert!(iter, local_state; region_kwargs(insert!, iter)...)
```

Comment:
Would it be crazy to store the `local_state` in the iterator/problem? Then these lines could be written as:
```julia
    extract!(iter; region_kwargs(extract!, iter)...)
    update!(iter; region_kwargs(update!, iter)...)
    insert!(iter; region_kwargs(insert!, iter)...)
```
So the reasoning would be:
1. It makes those lines simpler and more general.
2. The `local_state` may be useful to access anyway, say if you stop DMRG at a certain step and want to restart it, analyze the 2-site wavefunction in a callback, etc.

@emstoudenmire curious to hear your thoughts on that.